File: src/main/java/io/atesfactory/evrl/parser/Parser.java
package io.atesfactory.evrl.parser;

import io.atesfactory.evrl.loader.LoaderConfig;
import io.atesfactory.evrl.transformer.TransformerConfig;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Parser {
    public Config parse(String location) throws ParserException {
        Pattern pattern = Pattern.compile("evrl://(.*)(?=/[A-Za-z0-9]+:)/(.*)");
        Matcher matcher = pattern.matcher(location);

        if (!matcher.find()) {
            throw new ParserException("Could not extract transformers and loader");
        }

        ArrayList<TransformerConfig> transformers = new ArrayList<>(Arrays.asList(parseTransformers(matcher.group(1))));
        LoaderConfig loader = parseLoaderConfig(matcher.group(2));

        return new Config(transformers, loader);
    }

    private LoaderConfig parseLoaderConfig(String contentString) {
        String loaderName = contentString.substring(0, contentString.indexOf(":"));
        String loadable = contentString.substring(loaderName.length() + 1);

        return new LoaderConfig(loaderName, loadable);
    }

    private TransformerConfig[] parseTransformers(String transformerString) {
        if (transformerString == null || transformerString.isEmpty()) {
            return new TransformerConfig[0];
        }

        List<TransformerConfig> transformers = new ArrayList<>();
        for (String transformerName : transformerString.split("\\|")) {
            if (transformerName.contains(":")) {
                String[] transformerParts = transformerName.split(":");
                transformers.add(new TransformerConfig(transformerParts[0], Arrays.asList(transformerParts[1].split(","))));
                continue;
            }

            if (!transformerName.trim().isEmpty()) {
                transformers.add(new TransformerConfig(transformerName));
            }
        }

        return transformers.toArray(new TransformerConfig[0]);
    }
}

File: src/main/kotlin/io/atesfactory/evrl/Parser.kt
package io.atesfactory.evrl

data class Config(val transformerConfigs: ArrayList<TransformerConfig>, val loaderConfig: LoaderConfig)
data class TransformerConfig(val transformerName: String, val config: List<String> = emptyList())
data class LoaderConfig(val loaderName: String, val loadable: String)

class Parser {
    fun parse(location: String) : Config {
        val matchResult = Regex("evrl://(.*)(?=/[A-Za-z0-9]+:)/(.*)").find(location)
        val groups = matchResult?.groupValues ?: throw ParserException("Could not extract transformers and loader")
        val transformers = parseTransformers(groups[1])
        val loader = parseLoaderConfig(groups[2])

        return Config(arrayListOf(*transformers), loader)
    }

    private fun parseLoaderConfig(contentString: String): LoaderConfig {
        val loaderName = contentString.substringBefore(":")
        val loadable = contentString.removePrefix("$loaderName:")

        return LoaderConfig(loaderName, loadable)
    }

    private fun parseTransformers(transformerString: String?) : Array<TransformerConfig> {
        transformerString ?: return emptyArray()

        val transformers = ArrayList<TransformerConfig>()
        for (transformerName in transformerString.split("|")) {
            if(transformerName.contains(":")) {
                val transformerParts = transformerName.split(":")
                transformers.add(
                    TransformerConfig(transformerParts[0], transformerParts[1].split(","))
                )
                continue
            }

            if(transformerName.trim() != "") {
                transformers.add(TransformerConfig(transformerName))
            }
        }

        return transformers.toTypedArray()
    }
}

class ParserException(message: String) : RuntimeException(message)
